### ISSUE TITLE
0.4.0 Avalonia Update has been released!

### ISSUE THREAD

**Issue by Happypig375** [verybadcat/CSharpMath#112]
With the co[DEMONETIZED]us🦠 shutting down schools, I have had more time to work on CSharpMath lol :rocket:
## New additions
1. A massive refactor and rename across CSharpMath (left: 0.3.0, right: 0.4.0)
    - 4x test increase!
![image](https://user-images.githubusercontent.com/19922066/78969425-7b49ae00-7b39-11ea-9c21-27331c652838.png)
2. A new Avalonia front end! (Thanks @jp2masa!)
![Avalonia](https://user-images.githubusercontent.com/19922066/78373612-692db400-75fd-11ea-89c3-2f2a4f47784a.png)
3. CSharpMath.Editor improvements
   - Enhanced slash (#109)
   - A blinking cursor
   - Inner atom editing
   - Large operator with limits editing
![Editor](https://i.imgur.com/jokOjsU.gif)
4. New math mode symbol commands!
https://github.com/verybadcat/CSharpMath/blob/de1458afa2a2129cc928731885deb634205badc9/CSharpMath/Atom/LaTeXSettings.cs#L588-L818
5. `\colorbox`es!
![image](https://user-images.githubusercontent.com/19922066/78970707-89e59480-7b3c-11ea-8f46-ec4c8a6fde98.png)
6. More helpful error messages!
![image](https://user-images.githubusercontent.com/19922066/78970751-a5e93600-7b3c-11ea-9929-f6018019344c.png)
7. MathButton and TextButton!
   - Usage:
https://github.com/verybadcat/CSharpMath/blob/824d0df67b52c71ffbea970f165ad60b35ba16d3/CSharpMath.Forms.Example/CSharpMath.Forms.Example/Controls/MathKeyboard.xaml#L159-L161
   - This control is Xamarin.Forms only as Avalonia buttons can have arbitrary content including MathViews
8. Built-in Cyrillic support!
![image](https://user-images.githubusercontent.com/19922066/78971131-77b82600-7b3d-11ea-8539-d90c0853c9be.png)
9. Bugs fixed: #37, #59, #77 
10. CI for CSharpMath
![image](https://user-images.githubusercontent.com/19922066/78971383-062ca780-7b3e-11ea-9439-db879068905d.png)
11. C# 8 nullable reference types enabled in the whole repo
12. Removed dependency on the unmaintained SkiaSharp.Extended (Also fixes #77)

## What's next?
Since the exams, although delayed, are still coming up, I will need to temporarily leave once again. However, there are already [a bunch of planned items for the next update](https://github.com/verybadcat/CSharpMath/blob/e7127b2a77a4f15d4b31a53bbdd62c1ed134f139/Idea%20bucket.txt) once I come back. Stay tuned!

## Leftovers from 0.2.0 checklist (see https://github.com/verybadcat/CSharpMath/issues/104)

Item | Comments
-|-
#4 Merge with wpf-math | After looking into wpf-math, it depends heavily on the WPF infrastructure and is hard to make cross-platform. In contrast, CSharpMath is built with cross-platform in mind upon creation. A most probable resolution would be CSharpMath supporting the non-WPF platforms while wpf-math stays on WPF to facilitate division of labor.
F# instead of C# | Not really feasible as rewriting an entire solution will be extremely time-consuming but for little benefit. A better way to introduce F# would be in an entirely new project - currently planned for 0.6.0
Refactor TextAtoms and MathAtoms to have common atoms | The layout mechanism for MathAtoms and TextAtoms are too different for both atom hierarchies to unify. Not doing it.
Refactor MathAtoms to get rid of MathAtomType and AtomCloner | Done. All atoms now have a Clone() method.
Unit tests (and picture comparisons) | Done. Added as CSharpMath.Rendering.Tests
Tavis CI integration | Done. Added as GitHub Actions
An intuitive README for getting started | Will be done along with a new website. Currently planned for 0.6.0 - stay tuned!
Documentation on both NuGet and GitHub wiki | A website will be so much better. It will be much more maintainable and customizable. Probably won't do this.
Document release procedure | Done. Added as [ReleaseProcedure.txt](https://github.com/verybadcat/CSharpMath/blob/master/ReleaseProcedure.txt)
Promote awareness!! | Postponed to after 0.6.0 when the website is complete
Native platform support for the popular ones | Maybe. Probably won't do every platform as utilizing CI to automatically run unit tests will be difficult.
Implement `\middle` | Potential 0.5.0 candidate
Optimize loading times of the Android example project | Potential 0.5.0 candidate

## Stay safe🦺, 💪stay healthy🧼 and 🥚🐇Happy Easter🐣🐰!

Supersedes #104.

**Comment by ImadAbuDev** [comment]
Hi thanks for the update I update my Xamarin Forms Solution to 0.4.0.
Now I get an arrow for the function InitializeComponent(): 'Could not set up parent class, due to: Generic type definition failed to init, due to: VTable setup of type CSharpMath.Forms.BaseView2[TPainter,TContent] failed assembly:/storage/emulated/0/Android/data/com.companyname.mathx/files/.__override__/CSharpMath.Forms.dll type:BaseView2 member:(null) assembly:/storage/emulated/0/Android/data/com.companyname.mathx/files/.__override__/CSharpMath.Forms.dll type:BaseView`2 member:(null)'

**Comment by Happypig375** [comment]
try deleting bin/obj folders and rebuilding?

**Comment by ImadAbuDev** [comment]
tried it still not working also deleted nuget cache

**Comment by ImadAbuDev** [comment]
When I reset to 0.3.0 it works for me I don't know if the issue is from my solution or the update

**Comment by Happypig375** [comment]
I can investigate if you send me your solution.

**Comment by ImadAbuDev** [comment]
Sure check it https://we.tl/t-QoEgR1xKQL

**Comment by Happypig375** [comment]
Figured it out. Somehow CSharpMath.Forms 0.4.0 had a dependency on CSharpMath.SkiaSharp 0.4.0-beta instead of 0.4.0.
I have no idea how this happened.

**Comment by ImadAbuDev** [comment]
Any suggestions to fix this? So the bug isn't from my solution?

**Comment by Happypig375** [comment]
Pushed a 0.4.0.1.

**Comment by Happypig375** [comment]
Or you can explicitly specify a dependency on CSharpMath.SkiaSharp 0.4.0.

**Comment by Happypig375** [comment]
I guess I have to add a test for loading and using the latest NuGet.

**Comment by charlesroddie** [comment]
@Happypig375  has there been a change to mathPainter.Draw or to the definition of .Measure?

I am guessing that Draw was changed so that the point to be drawn is positioned at left middle of the object's rectangle, instead of the top left (https://github.com/verybadcat/CSharpMath/issues/12#issuecomment-405214055). In any case the API has changed so I will need to work out how to fix all usages of Draw, assuming that the new API is stable.

**Comment by charlesroddie** [comment]
It's looking like a change in the definition of Measure. I remember there were problems with the vertic
We have the following code to adjust y positioning based on whether you want the lining to be at y or the center or the top:

Only YTextAlign.Line works now so I suspect there is a bug with both 

```fsharp
let yCoordinateToDrawTo =
    y +
    match yAlign with
    | YTextAlign.Line -> 0.f // works
    | YTextAlign.Center -> -0.5f * (measure.Top + measure.Bottom) // fails
    | YTextAlign.Top -> - measure.Top // fails
```

**Comment by Happypig375** [comment]
The behavior of `Measure` has now been fixed and documented at https://github.com/verybadcat/CSharpMath/blob/6678311f2f4f711ce6992e20ea31cd095205f651/CSharpMath.Rendering.Tests/TestMeasure.cs#L40-L58

**Comment by charlesroddie** [comment]
I don't think that doc makes anything clear.

The previous behaviour is as follows: https://github.com/verybadcat/CSharpMath/issues/15#issuecomment-411877916

> If you draw at (x,y), (x,y) is the point at the left of the baseline. The y-coordinate of the baseline is y, of the top is y + Top and the coordinate of the bottom is y + Bottom. It's clear, and also standard. E.g. SkiaSharp text drawing positions the text from the baseline.

I would recommend reverting to this. If you want to discuss before changing, then can you give a definition that applies to the current Measure?

**Comment by charlesroddie** [comment]
Fix works. Thanks!

**Comment by angelklim** [comment]
Have been noticing missing text related to starting TeX with $$...$$.
For example, "$$p(x)=ax^2+bx+c, \text{ where } a \neq 0$$ A quadratic is a $2^{nd}$ degree polynomial." would only show "A quadratic is a 2nd degree polynomial." 
And "$$x$$ $$y$$" will only show "y"

Just $$...$$ by itself or after regular seems to display correctly.

**Comment by Happypig375** [comment]
Fixed in v0.4.2
Added as test: https://github.com/verybadcat/CSharpMath/blob/a3d657815523684da10d096a8e04a9eeeaefc9b5/CSharpMath.Rendering.Tests/Text/QuadraticPolynomial.png

**Comment by Happypig375** [comment]
Superseded by #190.
